pilot/test/integration/cloudfoundry/copilot_test.go

```diff
@@ -0,0 +1,277 @@
+
+	testState := newTestState(copilotAddr)
+	defer testState.tearDown()
+	copilotTLSConfig := testState.creds.ServerTLSConfig()
+
+	quitCopilotServer := make(chan struct{})
+
+	testState.addCleanupTask(func() { close(quitCopilotServer) })
+
+	t.Log("starting mock copilot grpc server...")
+
+	mockCopilot, err := bootMockCopilotInBackground(copilotAddr, copilotTLSConfig, quitCopilotServer)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	mockCopilot.PopulateRoute("abc.com", "127.0.0.1", appPort)
+
+	err = testState.config.Save(testState.configFilePath)
+	g.Expect(err).To(gomega.BeNil())
+
+	t.Log("building pilot...")
+
+	pilotSession, err := runPilot(testState.configFilePath, pilotPort)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	testState.addCleanupTask(func() {
+		pilotSession.Terminate()
+		g.Eventually(pilotSession, "5s").Should(gexec.Exit())
+	})
+
+	t.Log("checking if pilot ready")
+	g.Eventually(pilotSession.Out, "10s").Should(gbytes.Say(`READY`))
+
+	t.Log("checking if pilot received routes from copilot")
+	g.Eventually(func() (string, error) {
+		return curlPilot(registrationRoute)
+	}).Should(gomega.ContainSubstring("abc.com"))
+
+	g.Eventually(func() (string, error) {
+		return curlPilot(listenersRoute)
+	}).Should(gomega.ContainSubstring("http_connection_manager"))
+
+	t.Log("run envoy...")
+
+	err = testState.runEnvoy(pilotAddr)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	t.Log("curling the app with expected host header")
+
+	g.Eventually(func() error {
+		respData, err := curlApp(appEnvoyListener, "abc.com")
+		if err != nil {
+			return err
+		}
+		if respData != "hello" {
+			return fmt.Errorf("unexpected response data: %s", respData)
+		}
+		return nil
+	}, "5s", "100ms").Should(gomega.Succeed())
+}
+
+type testState struct {
+	// generated credentials for server and client
+	creds testhelpers.MTLSCredentials
+
+	// path on disk to store the config.yaml
+	configFilePath string
+
+	config *cloudfoundry.Config
+
+	cleanupTasks []func()
+	mutex        sync.Mutex
+}
+
+func (testState *testState) addCleanupTask(task func()) {
+	testState.mutex.Lock()
+	defer testState.mutex.Unlock()
+	testState.cleanupTasks = append(testState.cleanupTasks, task)
+}
+
+func newTestState(mockServerAddress string) *testState {
+	creds := testhelpers.GenerateMTLS()
+	clientTLSFiles := creds.CreateClientTLSFiles()
+	return &testState{
+		creds:          creds,
+		configFilePath: filepath.Join(creds.TempDir, "config.yaml"),
+		config: &cloudfoundry.Config{
+			Copilot: cloudfoundry.CopilotConfig{
+				ServerCACertPath: clientTLSFiles.ServerCA,
+				ClientCertPath:   clientTLSFiles.ClientCert,
+				ClientKeyPath:    clientTLSFiles.ClientKey,
+				Address:          mockServerAddress,
+				PollInterval:     10 * time.Second,
+			},
+			Envoy: cloudfoundry.EnvoyConfig{
+				Port: 6666,
+			},
+		},
+	}
+}
+
+func (testState *testState) runEnvoy(discoveryAddr string) error {
+	config := model.DefaultProxyConfig()
+	dir := os.Getenv("ISTIO_BIN")
```

Comment:
nice! thanks for sticking with the env var

Reply:
+1